### PR TITLE
fix: check unencrypted volumes from launch template, launch configuration and instance

### LINKED ISSUE #1167: bug: EC2EBSVolumeEncrypted rule fails to return non-compliance on unencrypted volumes
### What is the problem?

The rule `EC2EBSVolumeEncrypted` does not return non-compliance for at least some instances of unencrypted EBS volumes.

### Reproduction Steps

The following stack creates:
1. a standalone `BlockDevice` that is unencrypted
2. an autoscaling group that contains EC2 instances with unencrypted EBS volumes

To reproduce:
1. run `cdk synth`
2. observe there are *no* errors for unencrypted volumes

```python

from constructs import Construct
import aws_cdk as cdk
from aws_cdk import (
    Duration,
    Stack,
    aws_iam as iam,
    aws_ec2 as ec2,
    aws_autoscaling as autoscaling,
)
from cdk_nag import AwsSolutionsChecks


class PythonCdkExampleStack(Stack):

    def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)

        # VPC
        vpc = ec2.Vpc(self, "VPC",
            nat_gateways=0,
            subnet_configuration=[ec2.SubnetConfiguration(name="public",subnet_type=ec2.SubnetType.PUBLIC)]
            )

        # AMI
        amzn_linux = ec2.MachineImage.latest_amazon_linux(
            generation=ec2.AmazonLinuxGeneration.AMAZON_LINUX_2,
            edition=ec2.AmazonLinuxEdition.STANDARD,
            virtualization=ec2.AmazonLinuxVirt.HVM,
            storage=ec2.AmazonLinuxStorage.GENERAL_PURPOSE
            )

        # Instance Role and SSM Managed Policy
        role = iam.Role(self, "InstanceSSM", assumed_by=iam.ServicePrincipal("ec2.amazonaws.com"))
        role.add_managed_policy(iam.ManagedPolicy.from_aws_managed_policy_name("AmazonSSMManagedInstanceCore"))

        asg = autoscaling.AutoScalingGroup(self, "ExampleASG",
                                            vpc=vpc,
                                            role=role,
                                            instance_type=ec2.InstanceType("t3.micro"),
                                            machine_image=amzn_linux,
                                            min_capacity=1,
                                            max_capacity=1,
                                            require_imdsv2=True,
                                            block_devices=[
                                                autoscaling.BlockDevice(
                                                    device_name="/dev/xvda",
                                                    volume=autoscaling.BlockDeviceVolume.ebs(
                                                        volume_type=autoscaling.EbsDeviceVolumeType.GP2,
                                                        volume_size=50,
                                                        encrypted=False,
                                                        delete_on_termination=True
                                                    )),
                                                autoscaling.BlockDevice(
                                                        device_name="/dev/sdf",
                                                        volume=autoscaling.BlockDeviceVolume.ebs(
                                                            volume_size=20,
                                                            encrypted=False,
                                                            delete_on_termination=True
                                                    ))
                                                ]
                                            )

        bd = ec2.BlockDevice(device_name="example",
                            volume=ec2.BlockDeviceVolume.ebs(
                                volume_size=20,
                                encrypted=False,
                            ))


app = cdk.App()
PythonCdkExampleStack(app, "python-cdk-example")
cdk.Aspects.of(app).add(AwsSolutionsChecks(verbose=True))
app.synth()

```

### What did you expect to happen?

cdk-nag would return error for each unencrypted EBS volume, regardless of whether they have been declared explicitly, created as part of an EC2 instance, created as part of an autoscaling group, or by other means.

### What actually happened?

cdk-nag did not return any errors related to unencrypted EBS volumes.

### cdk-nag version

2.21.66

### Language

Python

### Other information

Tested with 2.21.8 as well. Same behavior.

### PR DESCRIPTION
**Pull request by clueleaf**
Fixes #1167

### CODE REVIEW

File: test/rules/EC2.test.ts [hunk 1]
```diff
@@ -784,20 +788,189 @@ describe('Amazon Elastic Compute Cloud (Amazon EC2)', () => {
...
         availabilityZone: stack.availabilityZones[0],
         size: Size.gibibytes(42),
         encrypted: false,
       });
       validateStack(stack, ruleId, TestType.NON_COMPLIANCE);
     });
-    test('Compliance', () => {
-      new Volume(stack, 'rVolume', {
+    test('Noncompliance 2 - CfnVolume', () => {
+      new CfnVolume(stack, 'rVolume', {
```
Comment (dontirun) [inline test/rules/EC2.test.ts:800]: Let's remove the `r` from the Ids ( ex. `rVolume` -> `Volume`). It's from a naming convention used for regular CloudFormation that I initially carried over to here but am removing as new PRs are made.

File: src/rules/ec2/EC2EBSVolumeEncrypted.ts [hunk 2]
```diff
@@ -19,6 +20,63 @@ export default Object.defineProperty(
...
+      } else {
+        const blockDeviceMappings = Stack.of(node).resolve(
+          launchTemplateData.blockDeviceMappings
+        );
+        for (const blockDeviceMapping of blockDeviceMappings) {
+          const encryption = NagRules.resolveIfPrimitive(
+            node,
+            blockDeviceMapping.ebs.encrypted
+          );
+          if (encryption != true) {
```
Comment (dontirun) [inline src/rules/ec2/EC2EBSVolumeEncrypted.ts:38]:
```suggestion
          if (encryption !== true) {
```

File: src/rules/ec2/EC2EBSVolumeEncrypted.ts [hunk 3]
```diff
@@ -19,6 +20,63 @@ export default Object.defineProperty(
...
+      );
+      if (blockDeviceMappings == undefined) {
+        return NagRuleCompliance.NON_COMPLIANT;
+      } else {
+        for (const blockDeviceMapping of blockDeviceMappings) {
+          const encryption = NagRules.resolveIfPrimitive(
+            node,
+            blockDeviceMapping.ebs.encrypted
+          );
+          if (encryption != true) {
```
Comment (dontirun) [inline src/rules/ec2/EC2EBSVolumeEncrypted.ts:56]:
```suggestion
          if (encryption !== true) {
```

File: src/rules/ec2/EC2EBSVolumeEncrypted.ts [hunk 4]
```diff
@@ -19,6 +20,63 @@ export default Object.defineProperty(
...
```
Comment (dontirun) [inline src/rules/ec2/EC2EBSVolumeEncrypted.ts:74]:
```suggestion
          if (encryption !== true) {
```

Comment (dontirun) [inline src/rules/ec2/EC2EBSVolumeEncrypted.ts:74]:
```suggestion
          if (encryption !== true) {
```

**Review by dontirun — CHANGES_REQUESTED** [review]
The[ rule `info`](https://github.com/cdklabs/cdk-nag/blob/14b7ad608e588ed66230500fb0fd822a30090ef8/src/packs/aws-solutions.ts#L260) should be updated to be a bit more helpful.

I think something like `The resource creates one or more EBS volumes that have encryption disabled.`

The EBS volume has encryption disabled

**Comment by clueleaf** [comment]
@dontirun 
Thank you for your review. I think your comments are all resolved now.

**Review by dontirun — APPROVED** [review]
Thanks for the contribution!

**Pull request merged**